design/baremetal-operator/sub-states.md

```diff
@@ -0,0 +1,185 @@
+
+* Provide access to the Ironic Node `provision_state` field.
+* Establish a direct one-to-one mapping between provision states and the
+  new field values.
+* Provide more granularity than what the Ironic API can report.
+* The values of the new field will not be a part of the API contract and can
+  change as a result of backend changes.
+* Define a non-empty detailed activity for each possible situation.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an operator reasonably familiar with the provisioning process, I would like
+to know what action is being executed on a BMH right now.
+
+## Design Details
+
+* The `ProvisionStatus` structure is extended with a new string field
+  `CurrentActivity` and a new integer field `Progress`.
+* The `Progress` value is a number from 0 to 100 that corresponds to estimated
+  percentage of the work that has been done so far.
+* The values of the fields are re-calculated based on the result described
+  below on each reconciliation.
+
+### Ironic steps basics
+
+This section explains the basics of Ironic clean/deploy steps. Feel free to
+skip it if you're familiar with Ironic.
+
+Ironic has three major processes that are of interest for us:
+
+* *Inspection* is a process of populating hardware information.
+* *Cleaning* is a process of preparing a node for being usable for deployment.
+* *Deployment* is a process of provisioning of a node.
+
+All three processes usually (with a few exceptions that are skipped for
+brevity) work by booting a ramdisk on the target node and executing actions
+from within it. When *fast track* mode is on, the ramdisk is booted only once
+and inspection, cleaning and deployment happen in a sequence.
+
+Cleaning and deployment processes are split into potentially pluggable *clean*
+or *deploy steps*. The said steps can be defined both inside Ironic itself or
+inside the ramdisk. When a process is running, the current step is stored in
+the Node's `clean_step` or `deploy_step` fields. All steps that are currently
+scheduled for execution are cached in
+`Node.driver_internal_info["clean_steps"]` or
+`Node.driver_internal_info["deploy_steps"]`.
+
+**Warning:** deploy (and in the future also automated clean) steps are
+populated on demand. Before the ramdisk is actually booted, not all steps may
+be in the cache.
+
+### Implementation Details/Notes/Constraints
+
+The new fields will be populated based on the following rules, where `Node` is
+an Ironic Node object. The rules are defined in pseudo-code.
+
+The calculations of the progress are based on the assumption that automated
+cleaning is enabled. It starts with the value of 0 and is updated based on the
+current detected activity.
+
+* IF `Node.provision_state` in "cleaning", "clean wait":
+
+  * IF `Node.clean_step` is null:
+
+    Activity is `Preparing to configure host`.
+
+  * IF `Node.clean_step["interface"]` == "raid":
+
+    Activity is `Configuring RAID`.
+
+  * IF `Node.clean_step["interface"]` == "bios":
```

Comment:
One issue we have with the patch to add RAID support is that the Node state returns to manageable after cleaning. To ensure we clean exactly once, the patch adds a new "Preparing" state (in which we kick off cleaning and immediately switch to the "Provisioning" state). It would be nice to either not have to add this new state, or to put all of the cleaning steps into the Preparing state - as it stands it's pretty misleading to the user. I don't think any changes to this proposal are required to do this, but it might be worth mentioning as one of the goals.